[Setup] Use WiX bootstrapper instead of a custom one

yuyoyuppe opened the pull request:
## Summary of the Pull Request

**What is this about:**
- remove custom bootstrapper and use wix's built-in one
- switch explorer addons to per user
- don't install VCM filter during setup
- bug fixes
- dotnet framework is now downloading using wix framework
**What is included in the PR:** 

**How does someone test / validate:** 
- build the installer and verify that VCM and file explorer modules still work
## Quality Checklist

- [ ] **Linked issue:** #xxx
- [ ] **Communication:** I've discussed this with core contributors in the issue. 
- [ ] **Tests:** Added/updated and all pass
- [ ] **Installer:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Docs:** Added/ updated
- [ ] **Binaries:** Any new files are added to WXS / YML
   - [ ] No new binaries
   - [ ] [YML for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/pipeline.user.windows.yml#L68) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries

## Contributor License Agreement (CLA)
A CLA must be signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/PowerToys) and sign the CLA.

yuyoyuppe committed:
[Setup] Use WiX bootstrapper instead of a custom one

DHowett commented on src/settings-ui/Settings.UI/App.xaml:
```diff
@@ -24,10 +24,10 @@
                 <Setter Property="Margin" Value="0,0,0,2" />
                 <Setter Property="Padding" Value="0,0,0,0" />
                 <Setter Property="HorizontalContentAlignment" Value="Stretch" />
-                <Setter Property="IsTabStop" Value="False"/>
```
Will this have an impact on accessibility or keyboard navigation? Or was this property a no-op.

stefansjfw replied:
this looks wrong, indeed

yuyoyuppe replied:
thanks, probably accidental change

DHowett commented on installer/PowerToysSetup/Product.wxs:
```diff
@@ -1,1321 +1,43 @@
-      
-    </ComponentGroup>
-  </Fragment>
+<Wix xmlns="http://schemas.microsoft.com/wix/2006/wi"
```
This is a minor complaint, but it would be much easier to review this changeset if you'd kept `Product.wxs` as the PowerToys MSI generator and introduced a _different_ `wxs` for the bootstrapper. Right now, comparing the actual changes made to the final MSI component is quite difficult due to the renames.

Is that something you'd be willing to change? Keeping `Product.wxs` and making the bootstrap one `Bootstrapper.wxs` or something?

DHowett replied:
(I'm pausing my review here; 45/51 files reviewed :smile:)

yuyoyuppe replied:
done :)

DHowett reviewed (changes requested):
Thanks so much for doing this! Glad to see vision become reality. Couple questions!

DHowett commented on installer/PowerToysSetup/PowerToysInstaller.wixproj:
```diff
@@ -1,90 +1,91 @@
+  <Import Project="..\..\src\Version.props" />
+  <PropertyGroup>
+    <DefineConstants>Version=$(Version)</DefineConstants>
+    <Name>PowerToysInstaller</Name>
```
This file got totally reformatted to change newlines; view the diff with "Hide Whitespace Changes" to see the actual two lines of change :smile:

DHowett commented on installer/PowerToysSetupCustomActions/CustomAction.cpp:
```diff
@@ -76,7 +75,7 @@ UINT __stdcall ApplyModulesRegistryChangeSetsCA(MSIHANDLE hInstall)
     ExitOnFailure(hr, "Failed to get installFolder.");
 
-    for (const auto& changeSet : getAllModulesChangeSets(installationFolder, false))
+    for (const auto& changeSet : getAllModulesChangeSets(installationFolder, true))
```
if nobody ever calls `getAllModulesChangeSets` with differnet values for this bool, should we just remove it?

htcfreek commented:
Does this have any impact on the existing installer parameters? Is it possible to run an admin install with special parameters or do I have to run the setup for every user on the computer? What is the default install directory/path?

crutkas commented:
@DHowett i flagged the default install DIR. this will get around UAC  for elevation but feels icky for a win32

DHowett commented:
> @DHowett i flagged the default install DIR. this will get around UAC  for elevation but feels icky for a win32

Hmm here's a Q: do we have to make the move to user dir at the same time as WiX Bundle? It seems fine if we do, we just have to drive down all the install differences.

crutkas commented:
Yes. This will be that tipping shift.

stefansjfw commented:
According to the code change, this should no longer be needed, right?
![image](https://user-images.githubusercontent.com/57057282/146756782-7533ecbe-4bd1-41dd-8c85-457198025e52.png)

stefansjfw commented:
Tested installer, PowerPreview and VCM seems to work correctly

stefansjfw commented:
not sure how important this is, but we lost the icon in `Add/Remove programs`
![image](https://user-images.githubusercontent.com/57057282/146759298-f71296b6-828c-44b8-8783-837f8a677915.png)

before:
![image](https://user-images.githubusercontent.com/57057282/146759342-96edf880-e7f1-447e-a317-56fda25928a9.png)

Aaron-Junker commented:
> not sure how important this is, but we lost the icon in `Add/Remove programs`
> ![image](https://user-images.githubusercontent.com/57057282/146759298-f71296b6-828c-44b8-8783-837f8a677915.png)
> 
> before:
> ![image](https://user-images.githubusercontent.com/57057282/146759342-96edf880-e7f1-447e-a317-56fda25928a9.png)
> 

I don't think that's high priority, but I'm sure people will complain about this.

SeraphimaZ commented:
Tested, VCM and PowerPreview work well.

yuyoyuppe commented:
> Does this have any impact on the existing installer parameters? Is it possible to run an admin install with special parameters or do I have to run the setup for every user on the computer? What is the default install directory/path?

Existing bootstrapper parameters will no longer work (btw, how do we change https://docs.microsoft.com/en-us/windows/powertoys/install#installer-arguments @crutkas), however WiX bootstrapper has a similar set of parameters. The most important one which couldn't be supported out of the box is `--extract_msi`, but this feature is supported by WiX tool called `dark`:

```cmd
"C:\Program Files (x86)\WiX Toolset v3.11\bin\dark.exe" -x msi S:\Programming\utilities\PowerToys\installer\PowerToysSetup\x64\Release\PowerToysSetup-0.53.1-x64.exe
```

Currently investigating the issue brought up by @stefansjfw.

DHowett reviewed (commented):
Alright, as of the merge with main you'll need to update release.yml in the bootstrapper phase :) this looks excellent though! Thanks!

yuyoyuppe commented:
Updated release.yml (I think it needs double-checking though!) and fixed missing icon issue :)

DHowett commented:
I've queued a test build on the Release pipeline -- here's hoping!

crutkas commented:
failure in setupcustomaction.
https://dev.azure.com/microsoft/Dart/_build/results?buildId=43009576&view=logs&j=f0810e3d-e629-5bb8-af19-7f15fc57dcc0&t=07d238a7-60a4-50ef-2d0f-bdda0083ff58

DHowett reviewed (commented):
More to Clint's point; I think that line 148 in release.yml needs updating if we aren't building the custom actions any longer. :)

DHowett commented:
Actually yeah, I'm confused about why this is an error. Huh. Did we change a project name?

crutkas commented:
that still exists in the buddy.yml  command: '.pipelines\build-installer-PTCustomActions.cmd' and was unchanged

yuyoyuppe commented:
We still need to build PTCustomActions separately to be able to sign in, right? Event though it's embedded into .msi installer. Logs show this error:
```log
error MSB4057: The target "PowerToysSetupCustomActions" does not exist in the project.
```

Still not sure what's causing this, but I've made `PowerToysSetupCustomActions`' project name to be more consistent.

crutkas commented:
1. Prompted a UAC.
2. 
```[8090:5764][2021-12-20T11:30:51]i001: Burn v3.11.2.4516, Windows v10.0 (Build 22526: Service Pack 0), path: C:\Users\crutkas\AppData\Local\Temp\{244F9054-292C-483A-A04E-330BA8FCFC47}\.cr\PowerToysSetup-0.0.1-x64.exe
[8090:5764][2021-12-20T11:30:51]i009: Command Line: '-burn.clean.room=C:\Users\crutkas\AppData\Local\Temp\MicrosoftEdgeDownloads\81de4abd-6f35-430e-8eda-c5d4840f66b2\PowerToysSetup-0.0.1-x64.exe -burn.filehandle.attached=764 -burn.filehandle.self=592'
[8090:5764][2021-12-20T11:30:51]i000: Setting string variable 'WixBundleOriginalSource' to value 'C:\Users\crutkas\AppData\Local\Temp\MicrosoftEdgeDownloads\81de4abd-6f35-430e-8eda-c5d4840f66b2\PowerToysSetup-0.0.1-x64.exe'
[8090:5764][2021-12-20T11:30:51]i000: Setting string variable 'WixBundleOriginalSourceFolder' to value 'C:\Users\crutkas\AppData\Local\Temp\MicrosoftEdgeDownloads\81de4abd-6f35-430e-8eda-c5d4840f66b2\'
[8090:5764][2021-12-20T11:30:51]i000: Setting string variable 'WixBundleLog' to value 'C:\Users\crutkas\AppData\Local\Temp\PowerToys_(Preview)_x64_20211220113051.log'
[8090:5764][2021-12-20T11:30:51]i000: Setting string variable 'WixBundleName' to value 'PowerToys (Preview) x64'
[8090:5764][2021-12-20T11:30:51]i000: Setting string variable 'WixBundleManufacturer' to value 'Microsoft Corporation'
[8090:63B0][2021-12-20T11:30:51]i000: Setting numeric variable 'WixStdBALanguageId' to value 1033
[8090:63B0][2021-12-20T11:30:51]i000: Setting version variable 'WixBundleFileVersion' to value '0.0.1.0'
[8090:5764][2021-12-20T11:30:51]i100: Detect begin, 2 packages
[8090:5764][2021-12-20T11:30:51]i000: Setting numeric variable 'HasDotnet3122' to value 1
[8090:5764][2021-12-20T11:30:51]i052: Condition 'HasDotnet3122' evaluates to true.
[8090:5764][2021-12-20T11:30:51]i101: Detected package: DotnetRuntime, state: Present, cached: None
[8090:5764][2021-12-20T11:30:51]i101: Detected package: PowerToysSetup_0.0.1_x64.msi, state: Absent, cached: None
[8090:5764][2021-12-20T11:30:51]i199: Detect complete, result: 0x0
[8090:63B0][2021-12-20T11:30:57]i000: Setting numeric variable 'EulaAcceptCheckbox' to value 1
[8090:5764][2021-12-20T11:30:57]i200: Plan begin, 2 packages, action: Install
[8090:5764][2021-12-20T11:30:57]w321: Skipping dependency registration on package with no dependency providers: DotnetRuntime
[8090:5764][2021-12-20T11:30:57]i000: Setting string variable 'WixBundleRollbackLog_PowerToysSetup_0.0.1_x64.msi' to value 'C:\Users\crutkas\AppData\Local\Temp\PowerToys_(Preview)_x64_20211220113051_000_PowerToysSetup_0.0.1_x64.msi_rollback.log'
[8090:5764][2021-12-20T11:30:57]i000: Setting string variable 'WixBundleLog_PowerToysSetup_0.0.1_x64.msi' to value 'C:\Users\crutkas\AppData\Local\Temp\PowerToys_(Preview)_x64_20211220113051_000_PowerToysSetup_0.0.1_x64.msi.log'
[8090:5764][2021-12-20T11:30:57]i201: Planned package: DotnetRuntime, state: Present, default requested: Present, ba requested: Present, execute: None, rollback: None, cache: No, uncache: No, dependency: None
[8090:5764][2021-12-20T11:30:57]i201: Planned package: PowerToysSetup_0.0.1_x64.msi, state: Absent, default requested: Present, ba requested: Present, execute: Install, rollback: Uninstall, cache: Yes, uncache: No, dependency: Register
[8090:5764][2021-12-20T11:30:57]i299: Plan complete, result: 0x0
[8090:5764][2021-12-20T11:30:57]i300: Apply begin
[8090:5764][2021-12-20T11:30:57]i010: Launching elevated engine process.
[8090:5764][2021-12-20T11:31:20]i011: Launched elevated engine process.
[8090:5764][2021-12-20T11:31:20]i012: Connected to elevated engine.
[42D0:5718][2021-12-20T11:31:21]i358: Pausing automatic updates.
[42D0:5718][2021-12-20T11:31:21]i359: Paused automatic updates.
[42D0:5718][2021-12-20T11:31:21]i360: Creating a system restore point.
[42D0:5718][2021-12-20T11:31:37]i361: Created a system restore point.
[42D0:5718][2021-12-20T11:31:37]i370: Session begin, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{59b11f12-87f8-4876-b11a-6a62fe56a21b}, options: 0x7, disable resume: No
[42D0:5718][2021-12-20T11:31:38]i000: Caching bundle from: 'C:\Users\crutkas\AppData\Local\Temp\{6536B340-F9F0-4383-93F1-77D0EFA52385}\.be\PowerToysSetup-0.0.1-x64.exe' to: 'C:\ProgramData\Package Cache\{59b11f12-87f8-4876-b11a-6a62fe56a21b}\PowerToysSetup-0.0.1-x64.exe'
[42D0:5718][2021-12-20T11:31:38]i320: Registering bundle dependency provider: {59b11f12-87f8-4876-b11a-6a62fe56a21b}, version: 0.0.1.0
[42D0:5718][2021-12-20T11:31:38]i371: Updating session, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{59b11f12-87f8-4876-b11a-6a62fe56a21b}, resume: Active, restart initiated: No, disable resume: No
[8090:8BFC][2021-12-20T11:31:38]i336: Acquiring container: WixAttachedContainer, copy from: C:\Users\crutkas\AppData\Local\Temp\MicrosoftEdgeDownloads\81de4abd-6f35-430e-8eda-c5d4840f66b2\PowerToysSetup-0.0.1-x64.exe
[8090:8BFC][2021-12-20T11:31:38]i000: Setting string variable 'WixBundleLastUsedSource' to value 'C:\Users\crutkas\AppData\Local\Temp\MicrosoftEdgeDownloads\81de4abd-6f35-430e-8eda-c5d4840f66b2\'
[8090:8C8C][2021-12-20T11:31:38]e000: Error 0x80070001: Failed to extract all files from container, erf: 1:2:0
[8090:8BFC][2021-12-20T11:31:38]e000: Error 0x80070001: Failed to wait for operation complete.
[8090:8BFC][2021-12-20T11:31:38]e000: Error 0x80070001: Failed to open container.
[8090:8BFC][2021-12-20T11:31:38]e000: Error 0x80070001: Failed to open container: WixAttachedContainer.
[8090:8BFC][2021-12-20T11:31:38]e312: Failed to extract payloads from container: WixAttachedContainer to working path: C:\Users\crutkas\AppData\Local\Temp\{6536B340-F9F0-4383-93F1-77D0EFA52385}\64C734420CFD72A9DA8FFEDE86EA41818D660714, error: 0x80070001.
[8090:5764][2021-12-20T11:31:38]e000: Error 0x80070001: Failed while caching, aborting execution.
[42D0:5718][2021-12-20T11:31:38]i372: Session end, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{59b11f12-87f8-4876-b11a-6a62fe56a21b}, resume: None, restart: None, disable resume: No
[42D0:5718][2021-12-20T11:31:38]i330: Removed bundle dependency provider: {59b11f12-87f8-4876-b11a-6a62fe56a21b}
[42D0:5718][2021-12-20T11:31:38]i352: Removing cached bundle: {59b11f12-87f8-4876-b11a-6a62fe56a21b}, from path: C:\ProgramData\Package Cache\{59b11f12-87f8-4876-b11a-6a62fe56a21b}\
[42D0:5718][2021-12-20T11:31:39]i371: Updating session, registration key: SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{59b11f12-87f8-4876-b11a-6a62fe56a21b}, resume: None, restart initiated: No, disable resume: No
[8090:5764][2021-12-20T11:31:40]i399: Apply complete, result: 0x80070001, restart: None, ba requested restart:  No
```

crutkas commented:
![image](https://user-images.githubusercontent.com/1462282/146823045-dfdf27e9-660f-404b-9444-84cb42065232.png)
"incorrect function"

DHowett commented:
@crutkas I think this is the thing I was telling you about. 

We need to produce the bundle, separate the engine from the engine, sign the engine, recombine the engine into the bundle, and sign the bundle. Like AppDriver. 

https://stackoverflow.com/questions/20381525/wix-digitally-sign-bootstrapper-project/20474482#20474482

We have to add steps to release.yml to call WiX insignia and sign the intermediates.

crutkas commented:
I'm going to branch this and test.

so based on steps via AppDriver.  
- Build MSI
- Sign MSI
- Build Bundle from MSI
- Extract engine from bundle (see block 1)
- Sign engine
- merge engine into bundle (see block 2)
- Sign final bundle

## block 1 - Extract engine

```
- script: '.\packages\WiX.3.14.0.4118\tools\insignia.exe -ib $(BUILD.BINARIESDIRECTORY)\$(BuildConfiguration)\$(BuildPlatform)\WindowsApplicationDriverBundle.exe -o $(BUILD.BINARIESDIRECTORY)\engine.exe'
  displayName: 'Insignia: Extract Engine from Bundle'
```

## block 2 - Merge engine

```
- script: '.\packages\WiX.3.14.0.4118\tools\insignia.exe -ab $(BUILD.BINARIESDIRECTORY)\engine.exe $(BUILD.BINARIESDIRECTORY)\$(BuildConfiguration)\$(BuildPlatform)\WindowsApplicationDriverBundle.exe -o $(BUILD.BINARIESDIRECTORY)\$(BuildConfiguration)\$(BuildPlatform)\WindowsApplicationDriverBundle.exe'
  displayName: 'Insignia: Merge Engine into Bundle'
```

crutkas commented:
testing via https://github.com/microsoft/PowerToys/pull/15085

crutkas commented:
1. Did a PR to your branch with fix for bootstrapper
2. Bootstrapper should be all someone would see, the MSI should just get everything passed into it
3. **User docs**, we'd create a PR against the docs team
4. I thought we were installing into the user dir now so we shouldn't need a UAC elevation, correct?  Or we need one more until 0.55?  (What happens if someone goes from 0.43 to .55 then for instance)

yuyoyuppe committed:
[Setup] Remove remaining elevation requirements except for installation folder

yuyoyuppe commented:
> Bootstrapper should be all someone would see, the MSI should just get everything passed into it

Yes, that's basically how it works already. We didn't have any special cmdline args handling in MSI, only in our custom bootstrapper.
> I thought we were installing into the user dir now so we shouldn't need a UAC elevation, correct? Or we need one more until 0.55? (What happens if someone goes from 0.43 to .55 then for instance)

I've tested and resolved 2 remaining elevation-requiring things in MSI with the latest commit. The only one remaining there is `Program Files` installation folder. However, we still need elevation to install dotnet runtime. I haven't found a way to conditionally require elevation with WiX depending on the package not being present in the system. 

Should we change installation folder to `%LOCALAPPDATA%`? Seems like it won't do much for us now since we still need to figure elevation requirement for dotnet dependency.

crutkas committed:
Crutkas wix bootstrapper fix (#1)

yuyoyuppe committed:
[Setup] Fully switch to per user installation

stefansjfw commented:
https://github.com/microsoft/PowerToys/issues/11777

Is this solved by this work?

DHowett commented:
> Yes, that's basically how it works already.

I think Clint is complaining that you pretty much have to go through two install screens; you're presented with the license twice, and the MSI presents its own UI. Since we're handling stuff in the bundle now we can probably make the MSI portion more silent.

crutkas commented:
What dustin said :)

yuyoyuppe committed:
[Setup] Format .wxs files

yuyoyuppe commented:
> What dustin said :)

Alright, that makes sense :) I've removed double MSI license dialog. Looks like there's no way to remove WiX Burn finish dialog though (the one at the end with "Close" button)

I've also changed installation dir to `%localappdata%` and made a complete switch to per-user install. Now our MSI installer truly doesn't need any elevation! See:

```xml
<Package InstallerVersion="200" Compressed="yes" InstallScope="perUser" InstallPrivileges="limited" 
```

Changing that required fighting some ICEs (mainly through adding `RemoveFolder` and `Registry` `KeyPath`s for every component). The only thing I've had to suppress is [ICE91](https://docs.microsoft.com/en-us/windows/win32/msi/ice91) for each and every file:
![image](https://user-images.githubusercontent.com/1828123/146991023-a804c713-248c-4174-88da-12362f9a7354.png)
Which I think doesn't make sense in our case, so can be safely ignored.

I've also taken a chance to reformat Product.wxs, since it accumulated a lot of mess over a long period of time, so for easier reviewing please refer to ebda64d commit.

yuyoyuppe commented:
The only thing requiring elevation during the whole installation process now is dotnet runtime dependency.

yuyoyuppe commented:
> #11777
Is this solved by this work?

Actually, I think I need to perform another round of testing now for cases like that and e.g. https://github.com/microsoft/PowerToys/issues/14725

crutkas commented:
I think let's not solve the UAC needed for 0.53 then due to the Runtime installer.  We will tackle that then in Jan.  I have an idea on how to do that.

1. Bootstrapper should be all that should be shown. 
2. Bootstrapper needs to prompt file path still
3. Silent call to MSI
4. Doc changes / removals from https://docs.microsoft.com/en-us/windows/powertoys/install#installer-arguments
4. Need to document how to extract MSI, now we're wix bundle, this should be well understood for an IT person

DHowett commented:
Lots of errors in the build now-

```
##[error]installer\PowerToysSetup\Product.wxs(1415,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1415): error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[error]installer\PowerToysSetup\Product.wxs(1421,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1421): error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[error]installer\PowerToysSetup\Product.wxs(1426,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1426): error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[error]installer\PowerToysSetup\Product.wxs(1431,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1431): error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[error]installer\PowerToysSetup\Product.wxs(1436,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1436): error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[error]installer\PowerToysSetup\Product.wxs(1441,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1441): error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[error]installer\PowerToysSetup\Product.wxs(1446,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1446): error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[error]installer\PowerToysSetup\Product.wxs(1451,0): Error CNDL0230: The Component/@Guid attribute's value '*' is not valid for this component because it does not meet the criteria for having an automatically generated guid. Components with registry keypaths and files cannot use an automatically generated guid. Create multiple components, each with one file and/or one registry value keypath, to use automatically generated guids.

```

crutkas commented:
I look at what ScreenToGif does and wonder why we don't do this?  It clearly shows dependencies and doesn't force UAC until needed.  Their current MSI lets you pick the folder then does a UAC if needed.

![image](https://user-images.githubusercontent.com/1462282/147010677-fc74ae88-1b53-4a6c-b267-0bac4fc4cb5e.png)

yuyoyuppe committed:
[Setup] add explicit Guids to all Components

yuyoyuppe commented:
![image](https://user-images.githubusercontent.com/1828123/147084800-e0901083-78fa-4344-abbf-05cfea9438b3.png)

@crutkas looks like ScreenToGif uses [WixSharp](https://github.com/oleg-shilo/wixsharp) toolkit which allows [customizing UI](https://github.com/oleg-shilo/wixsharp/blob/master/Source/src/WixSharp/Controls/CustomUI.cs) and further silencing/customizing UAC. Standard Bootstrapper doesn't seem to support those things.

@DHowett weird that I didn't get that locally, I wonder how internal CI differs in that regard 🤔. Added missing Guids to all components.

yuyoyuppe committed:
[Setup] Hide MSI UI

yuyoyuppe committed:
[Setup] Don't use UI Sequence for CA

yuyoyuppe committed:
[Setup] rebase fix

yuyoyuppe commented:
I've rebased the branch, also now install folder could be selected from WiX Bootstrapper UI, and we'll always create scheduled task for PowerToys process while MSI UI is hidden.

Also, I'm currently investigating an issue which seems to be caused by the newly added explicit `RemoveFolder`s to fix per-user install: if you choose a non-default install location, uninstallation won't delete any files. It only happens when uninstalled from WiX Bootstrapper (main user flow) or by using `msiexec` in silent mode. However, it works if uninstalled using full MSI UI. Logs indicate that `InstallFolder` is determined incorrectly in former case.

yuyoyuppe committed:
[Setup] Fix DetectPrevInstallPath CA order which prevented correctly determining folders to uninstall

DHowett commented:
Some of the language specific components are now emitting the GUID error; this is concerning as they are largely macro-generated :/

yuyoyuppe committed:
[Setup] Use generated GUIDs for macro-generated components

yuyoyuppe commented:
Modified the macro to generate GUIDs as well, we can run another build to see if it works.

yuyoyuppe commented:
Btw running the script provided by spelling bot no longer works:

```sh
sh-4.4$ bash spelling.bash
spelling.bash: line 34: jq: command not found
Can't open .github/actions/spelling/whitelist.txt: No such file or directory at -e line 9.
```

crutkas commented:
we shouldn't have a file called whitelist, it is expected.txt.  That is a non-inclusive term.

crutkas commented:
Kicked off a build.  https://dev.azure.com/microsoft/Dart/_build/results?buildId=43191045&view=results

yuyoyuppe commented:
@crutkas yes, I wonder if that message could be customized so it works out of the box

DHowett commented:
Not a bad idea, forcing their GUIDs. I was wondering why we couldn't go back to letting WiX automatically generate them -- we had to add them to *some* components for some reason, but did we need to do the languages too?

Probably better now that we're hardcoding them anyway...

crutkas commented:
so close.

Failed in Build MSI step.

```

     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1427): warning CNDL1118: The variable 'CompGUIDPrefix' with value 'EA375345-F52A-499B-A45E-108CE67EAF' was previously declared with value '85B4199F-5F89-42C3-B7C2-DBCD3E5267'. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[warning]installer\PowerToysSetup\Product.wxs(1430,0): Warning CNDL1118: The variable 'CompGUIDPrefix' with value 'F8EE7CE9-58EA-4850-B76F-C22362DC0E' was previously declared with value 'EA375345-F52A-499B-A45E-108CE67EAF'.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1430): warning CNDL1118: The variable 'CompGUIDPrefix' with value 'F8EE7CE9-58EA-4850-B76F-C22362DC0E' was previously declared with value 'EA375345-F52A-499B-A45E-108CE67EAF'. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
##[warning]installer\PowerToysSetup\Product.wxs(1433,0): Warning CNDL1118: The variable 'CompGUIDPrefix' with value 'F07F3210-79E2-4C38-AEF7-D8EBE06621' was previously declared with value 'F8EE7CE9-58EA-4850-B76F-C22362DC0E'.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(1433): warning CNDL1118: The variable 'CompGUIDPrefix' with value 'F07F3210-79E2-4C38-AEF7-D8EBE06621' was previously declared with value 'F8EE7CE9-58EA-4850-B76F-C22362DC0E'. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
       Link:
         C:\a\_work\1\s\installer\packages\WiX.3.11.2\build\..\tools\Light.exe -out C:\a\_work\1\s\installer\PowerToysSetup\x64\Release\PowerToysSetup-0.0.2-x64.msi -pdbout C:\a\_work\1\s\installer\PowerToysSetup\x64\Release\PowerToysSetup-0.0.2-x64.wixpdb -cultures:null -ext C:\a\_work\1\s\installer\packages\WiX.3.11.2\build\..\tools\\WixUtilExtension.dll -ext C:\a\_work\1\s\installer\packages\WiX.3.11.2\build\..\tools\\WixUIExtension.dll -ext C:\a\_work\1\s\installer\packages\WiX.3.11.2\build\..\tools\\WixNetFxExtension.dll -sacl -sice:ICE91 -sval -contentsfile obj\x64\Release\PowerToysInstaller.wixproj.BindContentsFileListnull.txt -outputsfile obj\x64\Release\PowerToysInstaller.wixproj.BindOutputsFileListnull.txt -builtoutputsfile obj\x64\Release\PowerToysInstaller.wixproj.BindBuiltOutputsFileListnull.txt -wixprojectfile C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj obj\x64\Release\pthC967EBBC18E63455754FDF3793A6A715\PTInstallDirDlg.wixobj obj\x64\Release\pthC967EBBC18E63455754FDF3793A6A715\PTLicenseDlg.wixobj obj\x64\Release\pthC967EBBC18E63455754FDF3793A6A715\WixUI_PTInstallDir.wixobj obj\x64\Release\Product.wixobj
         Windows Installer XML Toolset Linker version 3.11.2.4516
         Copyright (c) .NET Foundation and contributors. All rights reserved.
         
##[error]installer\PowerToysSetup\Product.wxs(633,0): Error LGHT0130: The primary key 'Remove_WindowsSettingsLanguagesFolder' is duplicated in table 'RemoveFile'.  Please remove one of the entries or rename a part of the primary key to avoid the collision.
     3>C:\a\_work\1\s\installer\PowerToysSetup\Product.wxs(633): error LGHT0130: The primary key 'Remove_WindowsSettingsLanguagesFolder' is duplicated in table 'RemoveFile'.  Please remove one of the entries or rename a part of the primary key to avoid the collision. [C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj]
     3>Done Building Project "C:\a\_work\1\s\installer\PowerToysSetup\PowerToysInstaller.wixproj" (default targets) -- FAILED.
     1>Done Building Project "C:\a\_work\1\s\installer\PowerToysSetup.sln" (Clean;PowerToysInstaller target(s)) -- FAILED.
Deferred Messages

```

yuyoyuppe commented:
@DHowett Switching to perUser install triggered a lot of stuff which wasn't required for perMachine - now **every** component requires explicit GUID, explicit `RemoveFolder` etc. :)

crutkas commented:
@yuyoyuppe if you need another build, just ping the thread.  I'll try and pay attention for the next 4 to 5 hours.

yuyoyuppe committed:
[Setup] Fix CI

yuyoyuppe commented:
Exactly the time I was asleep :) Pushed another fix.

DHowett commented:
Build queued! :)

DHowett commented:
Duplicate key, `Remove_RegistryPluginFolder`

yuyoyuppe committed:
[Setup] Fixed duplicate key ID

yuyoyuppe commented:
@DHowett fixed!

DHowett commented:
Remove_WindowsSettingsLanguageFolder :)

DHowett commented:
Really annoying how WIX doesn't dump all the errors at once ...

crutkas commented:
@yuyoyuppe any chance to get the last key dustin flagged?

yuyoyuppe committed:
[Setup] Another CI fix

yuyoyuppe commented:
@crutkas fixed

yuyoyuppe commented:
I believe this comment should make it green - I've made a script this time to replicate CI environment (should've done it a long time ago tbh🙂) and tested the build locally. It was actually pretty simple:

```batch
cd S:\Programming\utilities\PowerToys\.pipelines
S:
build-installer.cmd > log.txt
cat log.txt | perl -n -e"/error LGHT0103: The system cannot find the file '([^']+)'/ && print \"$1\n\"" | xargs -I[] -d'\n' dirname "[]" | xargs -I[] -d'\n' mkdir -p "[]"
cat log.txt | perl -n -e"/error LGHT0103: The system cannot find the file '([^']+)'/ && print \"$1\n\"" | xargs -I[] -d'\n' touch "[]"
```

crutkas commented:
kicked off a build.

We should kick off a new issue on cleaning up old batch files.  I would have removed that file without thinking

crutkas commented:
The installer ... installed!!!  

1. Context menu items don't seem to want to execute.  Tried Both Rename and Image Resizer
   1. Preview pane SVG and markdown did work
2. SVG thumbnails didn't seem to render and has edge icon (I didn't restart).  PDF thumbnail did

things we need to test
1. Would love to do a 0.53.2 test build to test upgrade paths

yuyoyuppe commented:
> Would love to do a 0.53.2 test build to test upgrade paths

Did some tests for the latest build, steps performed:

0.51->0.53 update flow (current -> next)
- build main branch as v0.51 with release endpoint pointing to my fork
- build current branch as v0.53, release in my fork
- install v0.51, try to update it to v0.53
- Stuck with this error: 
![image](https://user-images.githubusercontent.com/1828123/147606607-023485e5-8c88-4e52-91c0-ef3fc3051410.png)

because now the installer doesn't have access to Program Files. Unless we want to force the majority of users to uninstall manually, we need to fix it somehow.

I've tried the simplest way, changing

```xml
<Package InstallerVersion="200" Compressed="yes" InstallScope="perUser" InstallPrivileges="limited" Platform="x64" />
```

to

``xml
       ​<Package InstallerVersion="200" Compressed="yes" InstallScope="perMachine" InstallPrivileges="elevated" Platform="x64" />
```

and it worked. However, that just postpones the problem: I've tested 0.53(permachine)->0.55(peruser) update and it has the same issue. Essentially, we need to somehow explicitly invoke the uninstallation from an elevated WiX Bootstrapper instead of relying on MSI automatically triggering uninstallation when it detects the same `UpgradeCode`

> Context menu items don't seem to want to execute. Tried Both Rename and Image Resizer
Interesting, I cannot repro it on a clean VM, but I guess we need to fix the former issue before investigating this to have a final build.

franky920920 commented:
I bet there are serval docs that need to be updated. 

Since we're removing the bootstrapper, [the docs for the exe arguments](https://docs.microsoft.com/en-us/windows/powertoys/install#installer-arguments) should no longer exist.

crutkas commented:
This fails if i try and install create / manually install to `C:\Program Files\PowerToys`
![image](https://user-images.githubusercontent.com/1462282/147617088-5e7af2cb-f8a0-4858-a43e-0e0538abe044.png)

@DHowett my thought here for 0.53 we'll do a UAC style prompt at the start and default to `program files` so we have more time to work / test the edge cases like you installed to program files.

htcfreek commented:
When seeing the screenshot I personally miss the version number in the window title and app name (not processing text).

yuyoyuppe commented:
Ok, regarding the transition from perMachine to perUser. I've made a small prototype and it worked, so looks like we can leave .msi with limited privileges and it'll work as long as WiX Bootstrapper itself is elevated. What I did:

- compiled .exe which checks for installed PowerToys with version <=v0.51 (that means it's perMachine and requires elevation to remove) and uninstalls it
- added it as `ExePackage` which requires elevation (basically same as dotnet runtime)

That means .msi installer could become perUser starting from the next release and stay that way. Some downsides: 
- We still won't be able to install to program files. I haven't researched it much, but it looks like it'd be difficult to request elevation/UAC in case we detect some inaccessible installation path Program Files using WiX toolset.
- We will still need to run WiX Bootstrapper elevated. It seems like a sensible path for the foreseeable future considering dotnet runtime dependency though.
- We'll lose previous installation path

Thoughts?
cc: @crutkas @DHowett

crutkas commented:
@yuyoyuppe quick sync with @DHowett and we believe we should just do program files w/ UAC prompt for 0.53 and do a deeper dive for jan.  Lets do incremental changes.  We're already at a massive change currently in 0.53

yuyoyuppe committed:
[Setup] Use perMachine/elevated installation in msi

yuyoyuppe commented:
Alright, updated the branch and switched to perMachine. Also tested updating from 0.51 - it works.

crutkas commented:
kicked off dart build to do e2e testing.

crutkas commented:
One thing I noticed was when the app starts after install, it starts elevated

DHowett commented:
So, this is interesting! I tried to install this 0.0.2 build over 0.51! It "succeeded" instantly, but it didn't install over 0.51. The message indicating that we can't downgrade was never displayed... and I think the powertoys MSI was never launched!

On second launch, my only option was uninstall. It "uninstalled" something ... and I think it was also an empty uninstaller.

jaimecbernardo commented:
> So, this is interesting! I tried to install this 0.0.2 build over 0.51! It "succeeded" instantly, but it didn't install over 0.51. The message indicating that we can't downgrade was never displayed... and I think the powertoys MSI was never launched!
> 
> On second launch, my only option was uninstall. It "uninstalled" something ... and I think it was also an empty uninstaller.

It's installing a bundle of both the .net runtime and PowerToys. PowerToys already has a higher version so this isn't needed ... I think we might have to detect this manually.

https://stackoverflow.com/a/36209806/4594761

DHowett reviewed (changes requested):
So, I tested the new install path:

<img width="419" alt="image" src="https://user-images.githubusercontent.com/189190/147969370-5b044ad6-3a81-4722-8b09-2112f8a3cc74.png">


* It requires admin access,
* but it installs in the user's appdata folder,
* and shows up in all installed programs for all users

DHowett commented on installer/PowerToysSetup/PowerToys.wxs:
```diff
@@ -13,11 +13,17 @@
 
         <util:FileSearch Variable="HasDotnet3122" Path="[ProgramFiles64Folder]dotnet\shared\Microsoft.WindowsDesktop.App\3.1.22\System.Xaml.dll" Result="exists" />
 
+        <Variable Name="InstallFolder" Type="string" Value="[LocalAppDataFolder]PowerToys"/>
```
this is only true in the upgrade case! it shows one thing but then does another.

jaimecbernardo replied:
As discussed in https://github.com/microsoft/PowerToys/pull/15050#issuecomment-1002826076 , It's supposed to change back to installing in Program Files an be perMachine.
Seems all these changes were not made here yet. I'll apply those changes.

jaimecbernardo replied:
The upgrade scenario makes it so this value is ignored.
Will try to find a way to make the option hidden if we're upgrading.

jaimecbernardo replied:
Solved the upgrade to another folder scenario. Still need to populate this string, though.

jaimecbernardo committed:
Block downgrade scenarios

jaimecbernardo commented:
Added a commit to block the downgrade scenarios.

jaimecbernardo committed:
Fix uninstall of older version when upgrading

jaimecbernardo commented on installer/PowerToysSetup/Product.wxs:
```diff
@@ -173,1242 +192,1921 @@
+
+    <Fragment>
+        <Directory Id="TARGETDIR" Name="SourceDir">
+            <Directory Id="LocalAppDataFolder">
```
Changing this one back, as well, since we'll still be installing there by default, though I suspect this is not an issue.

jaimecbernardo committed:
Install in programs and revert to HKLM

jaimecbernardo commented:
> The installer ... installed!!!
> 
> 1. Context menu items don't seem to want to execute.  Tried Both Rename and Image Resizer
>    

I've reverted registry changes to HKLM instead of HKCU for the installer. At least PowerRename seems to need that when installed in Programs Folder.

Will take a closer look at the thumbnails/preview pane.

jaimecbernardo committed:
Fix spellchecker

jaimecbernardo committed:
Check Windows build number on bootstrapper

jaimecbernardo commented on installer/PowerToysSetup/PowerToys.wxs:
```diff
@@ -0,0 +1,54 @@
+              SuppressRepair="yes" />
+        </BootstrapperApplicationRef>
+
+        <util:FileSearch Variable="HasDotnet3122" Path="[ProgramFiles64Folder]dotnet\shared\Microsoft.WindowsDesktop.App\3.1.22\System.Xaml.dll" Result="exists" />
```
Added a commit to check this in the bootstrapper as well.

DHowett commented on installer/PowerToysSetup/PowerToys.wxs:
```diff
@@ -22,7 +22,7 @@
         <util:FileSearch Variable="HasDotnet3122" Path="[ProgramFiles64Folder]dotnet\shared\Microsoft.WindowsDesktop.App\3.1.22\System.Xaml.dll" Result="exists" />
 
-        <Variable Name="InstallFolder" Type="string" Value="[LocalAppDataFolder]PowerToys"/>
+        <Variable Name="InstallFolder" Type="string" Value="[ProgramFiles6432Folder]PowerToys"/>
```
(any particular reason why we chose 6432?)

htcfreek replied:
the old version is normal one (x64). imo a mistake

jaimecbernardo replied:
Seemed like the best default, from https://wixtoolset.org/documentation/manual/v3/bundle/bundle_built_in_variables.html
> ProgramFiles6432Folder - gets the well-known folder for CSIDL_PROGRAM_FILES on 64-bit Windows and CSIDL_PROGRAM_FILESX86 on 32-bit Windows.

But we don't support 32, so no issues with making it 64.

DHowett replied:
Oh, yeah, that seems great

jaimecbernardo replied:
Pushed a commit to fix it.

jaimecbernardo committed:
Use ProgramFiles64Folder as default install location

DHowett commented:
Alright! Let's fix up those conflicts and GO GO GO

jaimecbernardo commented:
> Alright! Let's fix up those conflicts and GO GO GO

@DHowett 

There are still some unsolved issues:

- The install folder option is shown on upgrade, but the value gets ignored. I couldn't find a way to fix this yet, still investigating. One option would be to leave it unfixed and just ignore it on upgrade for now. Other would be to show the original MSI UI instead, since that handles that case well enough, but it seems to have been decided against earlier.
- It seems there are some dangling files when updating as well. For example when trying to install on top of an older build it leaves files from the old version undeleted in the install folder. I still have to look closesly at this but it looks like the upgrade scenario is broken, currently.

jaimecbernardo commented:
> It seems there are some dangling files when updating as well. For example when trying to install on top of an older build it leaves files from the old version undeleted in the install folder. I still have to look closesly at this but it looks like the upgrade scenario is broken, currently.

Re-evaluated this. It no longer occurs. Must have fixed it by setting the keys to HKLM instead of HKCU instead. Behavior seems good now on upgrade.

jaimecbernardo committed:
Merge branch 'main' into pr15050

jaimecbernardo committed:
Fix spellchecker

jaimecbernardo committed:
Revert "Fix spellchecker"

This reverts commit f582b4d19e87cd22b7db1c471e30b89c02870b26.

jaimecbernardo committed:
fix spellcheck

jaimecbernardo committed:
Merge branch 'main' into wix_bootstrapper

jaimecbernardo committed:
Remove new keypath registry entries

jaimecbernardo committed:
Fix other folder upgrade scenario

jaimecbernardo committed:
fix spellchecker
